### PR TITLE
[SPARK-31448] [PYTHON] Fix storage level used in persist() in dataframe.py

### PR DESCRIPTION
**Pull request by abhishekd0907**
### What changes were proposed in this pull request?
Since the data is serialized on the Python side, we should make cache() in PySpark dataframes use StorageLevel.MEMORY_AND_DISK mode which has deserialized=false. This change was done to `pyspark/rdd.py` as part of SPARK-2014 but was missed from `pyspark/dataframe.py` 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Using existing tests

### CODE REVIEW

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -674,7 +674,7 @@ def cache(self):
         .. note:: The default storage level has changed to `MEMORY_AND_DISK` to match Scala in 2.0.
         """
         self.is_cached = True
-        self._jdf.cache()
+        self.persist(StorageLevel.MEMORY_AND_DISK)
```
Comment (srowen) [inline python/pyspark/sql/dataframe.py:677]: Hm, this should already be the default on the Scala side. I am not clear why this changes behavior.
I understand the idea but also not sure of all the implications here as so much has changed since SPARK-2014

Comment (abhishekd0907) [reply python/pyspark/sql/dataframe.py:677]: On scala side, default is `StorageLevel(true, true, false, true)` which has `deserialized=true`. It is getting used when scala `cache()` method is called directly as in the case with the current master branch.

However, since the data is already serialized on the Python side, Storage level to be used should have `deserialized=false`. With this code change persist will be called with `StorageLevel(True, True, False, False)` and hence deserialized will be set to false.

Comment (srowen) [reply python/pyspark/sql/dataframe.py:677]: I see, python StorageLevels are defined differently. Erm, @ScrapCodes or say @cloud-fan would you have an opinion on this? it does seem plausible but I don't know if there's a hidden reason for this difference.

Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py:677]: DataFrame itself is not serialized via Python unless you want to collect/createDataFrame/Python UDFs way whereas Python RDD are serialized via Python pickle. The code paths should be same as Scala's. I don't think this change is necessary.

Comment (cloud-fan) [reply python/pyspark/sql/dataframe.py:677]: yea, no matter you call `df.cache` in Scala or Python, the final RDD to cache is `RDD[CachedBatch]` not `RDD[Array[Byte]]`. I don't think SPARK-2014 applies here.

Comment (abhishekd0907) [reply python/pyspark/sql/dataframe.py:677]: Okay you might be right here but I observed a difference in the behavior of `pySparkDataframe.cache()` and `pySparkDataframe.persist()` because of differences in Storage Levels. When `cache()` is used, RDDs are spilled to disk but when `persist()` is used RDDs are persisted in memory. 

This is because `MemoryStore#putIteratorAsValues()` is used when `cache()` is called and `MemoryStore#putIteratorAsBytes()` is used when `persist()` is called due to difference in storage levels.

My example:
```
df = spark.range(1,4500000000).cache()
df.count()
AND
df = spark.range(1,4500000000).persist()
df.count()
```

I tried this on Spark version 2.4.6. I have attached RDD storage screenshots here. Correct me if I'm missing something and this is expected.
 
image


 
image

Comment (HyukjinKwon) [reply python/pyspark/sql/dataframe.py:677]: Does the same thing happen with Scala API too?

Comment (abhishekd0907) [reply python/pyspark/sql/dataframe.py:677]: No for scala behavior is the same whether `cache()` or `persist()` is used because storage level used is same in both the cases.

Comment (ScrapCodes) [reply python/pyspark/sql/dataframe.py:677]: For SPARK-2014, I cannot remember everything 6 years back. While I scrounge my memory, a bit more.

In python, turning on serialised should be okay, as everything is already serialized anyway.

Comment (ScrapCodes) [reply python/pyspark/sql/dataframe.py:677]: Isn't it true that every where(except Dataframe.py) whether scala or python APIs, persist means `MEMORY_ONLY` ?

Is there a specific reason to change that?

Comment (abhishekd0907) [reply python/pyspark/sql/dataframe.py:677]: No in `Dataset.scala` persist means persisting with default storage level `MEMORY_AND_DISK`
```
/**
   * Persist this Dataset with the default storage level (`MEMORY_AND_DISK`).
   *
   * @group basic
   * @since 1.6.0
   */
  def persist(): this.type = {
    sparkSession.sharedState.cacheManager.cacheQuery(this)
    this
  }

  /**
   * Persist this Dataset with the default storage level (`MEMORY_AND_DISK`).
   *
   * @group basic
   * @since 1.6.0
   */
  def cache(): this.type = persist()
```

Comment (abhishekd0907) [reply python/pyspark/sql/dataframe.py:677]: > Okay you might be right here but I observed a difference in the behavior of `pySparkDataframe.cache()` and `pySparkDataframe.persist()` because of differences in Storage Levels. When `cache()` is used, RDDs are spilled to disk but when `persist()` is used RDDs are persisted in memory.
> 
> This is because `MemoryStore#putIteratorAsValues()` is used when `cache()` is called and `MemoryStore#putIteratorAsBytes()` is used when `persist()` is called due to difference in storage levels.
> 
> My example:
> 
> ```
> df = spark.range(1,4500000000).cache()
> df.count()
> AND
> df = spark.range(1,4500000000).persist()
> df.count()
> ```
> 
> I tried this on Spark version 2.4.6. I have attached RDD storage screenshots here. Correct me if I'm missing something and this is expected.
> 
> image
> 
> image

@HyukjinKwon 
Can you please comment if the approach taken by this PR to solve this bug is correct or not? If it's not correct, can you please create a new JIRA and explain the issue and I can close this PR.
cc @srowen @cloud-fan @ScrapCodes

Comment (srowen) [reply python/pyspark/sql/dataframe.py:677]: Hm, yeah that is strange. What if Pyspark cache() just calls persist() in Pyspark, rather than cache() in Scala?

Comment (abhishekd0907) [reply python/pyspark/sql/dataframe.py:677]: @srowen 
Do you mean changing `self.persist(StorageLevel.MEMORY_AND_DISK)` to `self.persist()` at line 677? I believe the effect will be the same because the default storage level for Pyspark persist is also MEMORY_AND_DISK. 

```
@since(1.3)
    def persist(self, storageLevel=StorageLevel.MEMORY_AND_DISK):
        """Sets the storage level to persist the contents of the :class:`DataFrame` across
        operations after the first time it is computed. This can only be used to assign
        a new storage level if the :class:`DataFrame` does not have a storage level set yet.
        If no storage level is specified defaults to (`MEMORY_AND_DISK`).
        .. note:: The default storage level has changed to `MEMORY_AND_DISK` to match Scala in 2.0.
        """
        self.is_cached = True
        javaStorageLevel = self._sc._getJavaStorageLevel(storageLevel)
        self._jdf.persist(javaStorageLevel)
        return self
```

Comment (srowen) [reply python/pyspark/sql/dataframe.py:677]: Yes but that's the desired level, no? I understand there is a subtle difference between the meaning of that in Scala and Python, but, would that be the difference you are trying to fix here?

Comment (abhishekd0907) [reply python/pyspark/sql/dataframe.py:677]: using Pyspark's `self.persist(StorageLevel.MEMORY_AND_DISK)` or Pyspark's `self.persist()` will have the same impact since in both the cases Python's default storage level will be used i.e. `StorageLevel.MEMORY_AND_DISK = StorageLevel(True, True, False, False)` with `deserialized=false`. Hence both changes are equivalent and either will fix the bug.

Note that calling `self._sc._getJavaStorageLevel(StorageLevel.MEMORY_AND_DISK)` in pyspark doesn't return scala's default storage level. Instead, it returns `new StorageLevel(true, true, false, false)` which has `deserialized=false` and hence it will fix the bug.

However, in current master since scala's cache is called `self._jdf.cache()`, storage level uses is scala's default ie. `val MEMORY_AND_DISK = new StorageLevel(true, true, false, true)` which has `deserialized=true` leading to the bug.

Comment (srowen) [reply python/pyspark/sql/dataframe.py:677]: OK, right. That could be a simpler fix. But then I'm still not so much following why there is a behavior difference per https://github.com/apache/spark/pull/29242#discussion_r462086489 I'm trying to understand whether this difference is expected or not, per earlier comments.

Comment (abhishekd0907) [reply python/pyspark/sql/dataframe.py:677]: @srowen
The difference in behavior is due to this code in `BlockManager.scala`.  
```
// Put it in memory first, even if it also has useDisk set to true;
        // We will drop it to disk later if the memory store can't hold it.
        if (level.deserialized) {
          memoryStore.putIteratorAsValues(blockId, iterator(), classTag) match {
            case Right(s) =>
              size = s
            case Left(iter) =>
              // Not enough space to unroll this block; drop to disk if applicable
              if (level.useDisk) {
                logWarning(s"Persisting block $blockId to disk instead.")
                diskStore.put(blockId) { channel =>
                  val out = Channels.newOutputStream(channel)
                  serializerManager.dataSerializeStream(blockId, out, iter)(classTag)
                }
                size = diskStore.getSize(blockId)
              } else {
                iteratorFromFailedMemoryStorePut = Some(iter)
              }
          }
        } else { // !level.deserialized
          memoryStore.putIteratorAsBytes(blockId, iterator(), classTag, level.memoryMode) match {
            case Right(s) =>
              size = s
            case Left(partiallySerializedValues) =>
              // Not enough space to unroll this block; drop to disk if applicable
              if (level.useDisk) {
                logWarning(s"Persisting block $blockId to disk instead.")
                diskStore.put(blockId) { channel =>
                  val out = Channels.newOutputStream(channel)
                  partiallySerializedValues.finishWritingToStream(out)
                }
                size = diskStore.getSize(blockId)
              } else {
                iteratorFromFailedMemoryStorePut = Some(partiallySerializedValues.valuesIterator)
              }
          }
        }
```

When StorageLevel has `deserialized=true` `putIteratorAsValues` is used and when `deserialized=false`, `putIteratorAsBytes` is used. Since data in python is always serialized as mentioned by @ScrapCodes, we should always use `putIteratorAsBytes`. To do this, we need to always keep StorageLevel with `deserialized=false`. Calling pyspark's `persist()` ensures that.

@cloud-fan @HyukjinKwon Can you please comment if the approach taken by this PR is correct?

Comment (cloud-fan) [reply python/pyspark/sql/dataframe.py:677]: > Since data in python is always serialized

As I explained, this is not true in Spark SQL. The data to be cached is `RDD[CachedBatch]`. We should make sure `df.cache/persist` is consistent between python and scala/java.

Comment (abhishekd0907) [reply python/pyspark/sql/dataframe.py:677]: @cloud-fan @srowen 
Okay so, currently `scalaDataFrame.cache()` and `pySparkDataFrame.cache()` show consistent behavior. But `scalaDataFrame.persist()` and `pySparkDataFrame.persist()` are not consistent because `pysparkDataframe.persist()` uses `StorageLevel(true, true, false, false)` but `scalaDataframe.persist()` uses `StorageLevel(true, true, false, true)`. Should we change `pysparkDataframe.persist()` to directly call `self._jdf.persist()` ?

Comment (cloud-fan) [reply python/pyspark/sql/dataframe.py:677]: Yea I think so

Comment (abhishekd0907) [reply python/pyspark/sql/dataframe.py:677]: okay i'll update the PR

Comment (abhishekd0907) [reply python/pyspark/sql/dataframe.py:677]: @cloud-fan as you explained, data is not serialized when performing SQL operations with pyspark. 

So should we revert changes done as part of [SPARK-12091](https://issues.apache.org/jira/browse/SPARK-12091) since users may need to set storage levels which have deserialized is equal to true?

Comment (cloud-fan) [reply python/pyspark/sql/dataframe.py:677]: I'm not sure if it worth. The `CachedBatch` is mostly a wrapper of `byte[]` so keeping the serialized or deserialized form in memory doesn't matter too much.

I think the important thing is to keep Scala and Python `cache/persistent` consistent, although the Scala side has more builtin `StorageLevel`s.

**Comment by srowen** [comment]
Are you sure the last commit is consistent with the discussion above? I'm not quite following.

**Comment by abhishekd0907** [comment]
> Are you sure the last commit is consistent with the discussion above? I'm not quite following.

@srowen 
As @cloud-fan suggested in the comment below, the only change needed is to keep the storage level in `scalaDataFrame.persist()` and `pysparkDataframe.persist()` same i.e. `StorageLevel(true, true, false, true)`

> I'm not sure if it worth. The CachedBatch is mostly a wrapper of byte[] so keeping the serialized or deserialized form in memory doesn't matter too much.
I think the important thing is to keep Scala and Python cache/persistent consistent, although the Scala side has more builtin StorageLevels.

@cloud-fan let me know in case I misunderstood your point or you intend to achieve the same thing in some other way.
Apologies for so much delay in updating this PR.

File: python/pyspark/sql/dataframe.py [hunk 2]
```diff
@@ -678,13 +678,14 @@ def cache(self):
         return self
 
     @since(1.3)
-    def persist(self, storageLevel=StorageLevel.MEMORY_AND_DISK):
+    def persist(self, storageLevel=StorageLevel.MEMORY_AND_DISK_DESER):
```
Comment (cloud-fan) [inline python/pyspark/sql/dataframe.py:681]: Now I see the confusion. In Scala, `MEMORY_AND_DISK` means `deserialized=true`, while in Python, `MEMORY_AND_DISK` means `deserialized=false`.

File: python/pyspark/sql/dataframe.py [hunk 3]
```diff
@@ -678,13 +678,14 @@ def cache(self):
...
         """Sets the storage level to persist the contents of the :class:`DataFrame` across
         operations after the first time it is computed. This can only be used to assign
         a new storage level if the :class:`DataFrame` does not have a storage level set yet.
-        If no storage level is specified defaults to (`MEMORY_AND_DISK`).
+        If no storage level is specified defaults to (`MEMORY_AND_DISK_DESER`)
 
-        .. note:: The default storage level has changed to `MEMORY_AND_DISK` to match Scala in 2.0.
+        .. note:: The default storage level has changed to `MEMORY_AND_DISK_DESER` to match Scala
+            in 2.0.
```
Comment (cloud-fan) [inline python/pyspark/sql/dataframe.py:688]: `in 3.0`?

Comment (cloud-fan) [inline python/pyspark/sql/dataframe.py:688]: `in 3.0`?

File: python/pyspark/storagelevel.py [hunk 4]
```diff
@@ -56,3 +56,5 @@ def __str__(self):
 StorageLevel.MEMORY_AND_DISK = StorageLevel(True, True, False, False)
 StorageLevel.MEMORY_AND_DISK_2 = StorageLevel(True, True, False, False, 2)
 StorageLevel.OFF_HEAP = StorageLevel(True, True, True, False, 1)
+
```
Comment (HyukjinKwon) [inline python/pyspark/storagelevel.py:59]: Let's remove newline here.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by abhishekd0907** [comment]
@srowen @cloud-fan @HyukjinKwon
I've handled all the comments and all the tests are passing. Can you please review the PR further?

**Comment by srowen** [comment]
@cloud-fan looks right, just seeing if you can confirm

**Comment by HyukjinKwon** [comment]
I am okay with this change.

**Comment by ScrapCodes** [comment]
Looks good to me too!

**Comment by abhishekd0907** [comment]
@cloud-fan 
Other folks seem to be okay with the change. can you please review as well?

**Comment by srowen** [comment]
Merged to master
